src/abstractdataframe/sort.jl

```diff
@@ -109,24 +110,48 @@ function DFPerm(ords::AbstractVector{O}, cols::T) where {O<:Ordering, T<:Tuple}
     if length(ords) != length(cols)
         error("DFPerm: number of column orderings does not equal the number of columns")
     end
-    DFPerm{typeof(ords), T}(ords, cols)
+    DFPerm(Tuple(ords), cols)
 end
 
 DFPerm(o::Union{Ordering, AbstractVector}, df::AbstractDataFrame) =
     DFPerm(o, ntuple(i -> df[!, i], ncol(df)))
 
-# get ordering function for the i-th column used for ordering
-col_ordering(o::DFPerm{O}, i::Int) where {O<:Ordering} = o.ord
-col_ordering(o::DFPerm{V}, i::Int) where {V<:AbstractVector} = o.ord[i]
+@inline col_ordering(o::Ordering) = o
+@inline ord_tail(o::Ordering) = o
+@inline col_ordering(o::Tuple{Vararg{Ordering}}) = @inbounds o[1]
+@inline ord_tail(o::Tuple{Vararg{Ordering}}) = Base.tail(o)
+
+Sort.lt(o::DFPerm{<:Any, Tuple{}}, a, b) = false
+
+function Sort.lt(o::DFPerm{<:Any, <:Tuple}, a, b)
+    ord = o.ord
+    cols = o.cols
+    length(cols) > 16 && return  unstable_lt(ord, cols, a, b)
```

Comment:
Add a comment explaining how the 16 threshold was chosen?
```suggestion
    length(cols) > 16 && return unstable_lt(ord, cols, a, b)
```

Reply:
added. I have not tuned 16 specifically. I just assume that 16 is a safe threshold. Probably we could pass some higher number here, but I think that normally one does not sort on more than something like 4 columns.